libp2p-websocket/src/desktop.rs

```diff
@@ -0,0 +1,253 @@
+			// Need to suffix `/ws` to each client address.
+			let client_addr = client_addr
+				.encapsulate("/ws")
+				.expect("the /ws suffix is always valid");
+
+			// Upgrade the listener to websockets like the websockets library requires us to do.
+			let upgraded = stream.and_then(|stream| {
+				stream
+					.into_ws()
+					.map_err(|e| IoError::new(IoErrorKind::Other, e.3))
+					.and_then(|stream| {
+						// Accept the next incoming connection.
+						stream
+							.accept()
+							.map_err(|err| IoError::new(IoErrorKind::Other, err))
+							.map(|(client, _http_headers)| {
+								// Plug our own API on top of the `websockets` API.
+								let framed_data = client
+                                    .map_err(|err| IoError::new(IoErrorKind::Other, err))
+                                    .sink_map_err(|err| IoError::new(IoErrorKind::Other, err))
+                                    .with(|data| Ok(OwnedMessage::Binary(data)))
+                                    .and_then(|recv| {
+                                        match recv {
+                                            OwnedMessage::Binary(data) => Ok(Some(data)),
+                                            OwnedMessage::Text(data) => Ok(Some(data.into_bytes())),
+                                            OwnedMessage::Close(_) => Ok(None),
+                                            // TODO: pings and pongs ; freaking hard
+                                            _ => unimplemented!()
+                                        }
+                                    })
+                                    // TODO: is there a way to merge both lines into one?
+                                    .take_while(|v| Ok(v.is_some()))
+                                    .map(|v| v.expect("we only take while this is Some"));
+
+								let read_write = RwStreamSink::new(framed_data);
+								Box::new(read_write) as Box<AsyncStream>
+							})
+					})
+					.map(|s| Box::new(Ok(s).into_future()) as Box<Future<Item = _, Error = _>>)
+					.into_future()
+					.flatten()
+			});
+
+			(
+				Box::new(upgraded) as Box<Future<Item = _, Error = _>>,
+				client_addr,
+			)
+		});
+
+		Ok((listen, new_addr))
+	}
+
+	fn dial(self, original_addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)> {
+		let mut inner_addr = original_addr.clone();
+		match inner_addr.pop() {
+			Some(AddrComponent::WS) => {}
+			_ => return Err((self, original_addr)),
+		};
+
+		let inner_dial = match self.transport.dial(inner_addr) {
+			Ok(d) => d,
+			Err((transport, _)) => {
+				return Err((
+					WsConfig {
+						transport: transport,
+					},
+					original_addr,
+				));
+			}
+		};
+
+		let dial = inner_dial.into_future().and_then(|connec| {
+			// We pass a dummy address to `ClientBuilder` because it is never used anywhere
+			// in the negotiation anyway, and we use `async_connect_on` to pass a stream.
+			ClientBuilder::new("ws://127.0.0.1:80")
```

Comment:
Can we fork the WS library to make this cleaner or is it not worth it?

Reply:
I would say it's not worth it. Since it's a breaking change they would have to bump their version, leading to more burden.